dcs_core/integrations/databases/postgres.py

```
@@ -415,6 +422,53 @@ def _normalize_metrics(value):
+            if isinstance(distinct_count, (int, float)) and distinct_count < 20:
+                quoted = self.quote_column(col_name)
+
+                # JSON needs ::text for grouping
+                if dtype in ("json", "jsonb"):
+                    group_expr = f"{quoted}::text"
+                else:
+                    group_expr = quoted
+
+                dist_query = (
+                    f"SELECT {group_expr}, COUNT(*) "
+                    f"FROM {qualified_table} GROUP BY {group_expr} ORDER BY COUNT(*) DESC"
```

Ruff (0.14.6)

Possible SQL injection vector through string-based query construction

(S608)